src/rust/cryptography-x509-validation/src/types.rs

```diff
@@ -110,9 +112,144 @@ impl<'a> DNSPattern<'a> {
     }
 }
 
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum IPAddress {
+    V4(std::net::Ipv4Addr),
+    V6(std::net::Ipv6Addr),
+}
+
+/// An `IPAddress` represents an IP address as defined in [RFC 5280 4.2.1.6].
+///
+/// [RFC 5280 4.2.1.6]: https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.6
+impl IPAddress {
+    pub fn from_str(s: &str) -> Option<Self> {
+        std::net::IpAddr::from_str(s).ok().map(Self::from)
+    }
+
+    /// Constructs an `IPAddress` from a slice. The provided data must be
+    /// 4 (IPv4) or 16 (IPv6) bytes in "network byte order", as specified by
+    /// [RFC 5280].
+    ///
+    /// [RFC 5280]: https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.6
+    pub fn from_bytes(b: &[u8]) -> Option<Self> {
+        match b.len() {
+            4 => {
+                let b: [u8; 4] = b.try_into().ok()?;
+                Some(std::net::IpAddr::from(b).into())
+            }
+            16 => {
+                let b: [u8; 16] = b.try_into().ok()?;
+                Some(std::net::IpAddr::from(b).into())
+            }
+            _ => None,
+        }
+    }
+
+    /// Parses the octets of the `IPAddress` as a mask. If it is well-formed,
+    /// i.e., has only one contiguous block of set bits starting from the most
+    /// significant bit, a prefix is returned.
+    pub fn as_prefix(&self) -> Option<u8> {
+        let (leading, total) = match self {
+            Self::V4(a) => {
+                let data = u32::from_be_bytes(a.octets());
+                (data.leading_ones(), data.count_ones())
+            }
+            Self::V6(a) => {
+                let data = u128::from_be_bytes(a.octets());
+                (data.leading_ones(), data.count_ones())
+            }
+        };
+
+        if leading != total {
+            None
+        } else {
+            Some(leading as u8)
+        }
+    }
+
+    /// Returns a new `IPAddress` with the first `prefix` bits of the `IPAddress`.
+    ///
+    /// ```rust
+    /// # use cryptography_x509_validation::types::IPAddress;
+    /// let ip = IPAddress::from_str("192.0.2.1").unwrap();
+    /// assert_eq!(ip.mask(24), IPAddress::from_str("192.0.2.0").unwrap());
+    /// ```
+    pub fn mask(&self, prefix: u8) -> Self {
+        match self {
+            Self::V4(a) => {
+                let prefix = 32u8.checked_sub(prefix).unwrap_or(0).into();
+                let masked = u32::from_be_bytes(a.octets())
+                    & u32::MAX
+                        .checked_shr(prefix)
+                        .unwrap_or(0)
+                        .checked_shl(prefix)
+                        .unwrap_or(0);
```

Comment:
This is the same as `u32::MAX - (1 << prefix - 1)`, more or less, right? Just trying to make sure I understand the logic.

Reply:
And the checks are basically handling the case where prefix is 32, right?